python/lsst/ts/standardscripts/maintel/m1m3/check_actuators.py

```diff
@@ -0,0 +1,258 @@
+# for details of code ownership.
+#
+# This program is free software: you can redistribute it and/or modify
+# it under the terms of the GNU General Public License as published by
+# the Free Software Foundation, either version 3 of the License, or
+# (at your option) any later version.
+#
+# This program is distributed in the hope that it will be useful,
+# but WITHOUT ANY WARRANTY; without even the implied warranty of
+# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+# GNU General Public License for more details.
+#
+# You should have received a copy of the GNU General Public License
+# along with this program. If not, see <https://www.gnu.org/licenses/>.
+
+__all__ = ["CheckActuators"]
+
+
+import time
+import yaml
+import asyncio
+
+from lsst.ts.salobj import BaseScript
+from lsst.ts.idl.enums.Script import ScriptState
+from lsst.ts.idl.enums.MTM1M3 import DetailedState
+from lsst.ts.observatory.control.maintel.mtcs import MTCS, MTCSUsages
+
+from lsst.ts.cRIOpy.M1M3FATable import (
+    FATABLE,
+    FATABLE_ID,
+)
+
+
+class CheckActuators(BaseScript):
+    """Perform a M1M3 bump test on either a selection of individual
+    actuators or on all actuators..
+
+    Parameters
+    ----------
+    index : `int`
+        Index of Script SAL component.
+
+    Notes
+    -----
+    **Checkpoints**
+
+    - "Running bump test on FA ID: {id}.": Check individual actuator.
+    - "M1M3 bump test completed.": Check complete.
+
+    """
+
+    def __init__(self, index, add_remotes: bool = True):
+        super().__init__(index=index, descr="Bump Test on M1M3 Actuators")
+
+        self.mtcs = MTCS(
+            domain=self.domain,
+            intended_usage=None if add_remotes else MTCSUsages.DryTest,
+            log=self.log,
+        )
+
+        # Average duration (seconds) of a bump test on a single actuator
+        self.time_one_bump = 25
+
+        # Getting list of actuator ids from mtcs
+        self.m1m3_actuator_ids = self.mtcs.get_m1m3_actuator_ids()
+        self.m1m3_secondary_actuator_ids = self.mtcs.get_m1m3_actuator_secondary_ids()
+
+        # Actuators that will be effectively tested
+        self.actuators_to_test = self.m1m3_actuator_ids.copy()
+
+    async def assert_feasibility(self):
+        """Verify that the system is in a feasible state before
+        running bump test.
+        """
+
+        for comp in self.mtcs.components_attr:
+            if comp != "mtm1m3":
+                self.log.debug(f"Ignoring component {comp}.")
+                setattr(self.mtcs.check, comp, False)
+
+        # Check all enabled, liveliness and the m1m3 detailed state
+        await asyncio.gather(
+            self.mtcs.assert_all_enabled(),
+            self.mtcs.assert_liveliness(),
+            self.mtcs.assert_m1m3_detailed_state(
+                {
+                    DetailedState.PARKED,
+                    DetailedState.PARKEDENGINEERING,
+                    DetailedState.ACTIVE,
```

Comment:
You cannot (yet) run the bump test on the active mirror. Note T018 is a different beast. If we want that to run on the active mirror (and given the CSC can run that probably better, plus we will be able to run a bump test from EUI, which I am unable to do now), I need a ticket for M1M3.

You can still run T018-like bump test, but that will require you to implement T018-like logic, which is a lot more difficult than simply firing bump_test_FA.

I suggest leaving ACTIVE (and ACTIVEENGINEERING) out now. The problem with bump testing while active is that it will not tell you anything about whatever a FA is performing as expected - you need to raise the mirror first, and if an FA is misbehaving, it might be too late to save the mirror as you are rising it. Running bump test with mirror active is like using a tool to verify that a function (raising the mirror) can proceed after it already finished.

We now run T018 during mirror verification, but that's mainly to see FAs works at elevation. 002, the bump test while lowered, will do a better job. I don't envision why we shall run bump test when active during operation to verify FA performance.

Reply:
Hi Petr. One question here. This will only check the m1m3 state (raised or lowered), without changing its state. So, considering that we will use this script to run bump tests in both states (either for mirror lowered and raised, which will be done using the m1m3_lower and m1m3_raise), we should keep this checking, don't you think?

Reply:
Keep in only PARKED and PARKEDENGINEERING states. We will handle ACTIVE states later (if needed).